### PR TITLE
MC/CPU: optimized reduce-multi

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Optimizes CPU based reduction routine when multiple src buffers are passed (used in the collectives with RADIX > 2). In this case  we can unroll 1 loop of the reduction (taking 1 entry from each buffer with fixed stride).

## Why ?
Gives up to 10% overall collective improvement (see xls
[reduce_multi.xlsx](https://github.com/openucx/ucc/files/6540166/reduce_multi.xlsx)

## How ?
Explicit loop unroll for "size" (number of strided buffers) up to 7 (radix=8 case). For larger sizes (radixes) use unrolled 3 version and potentially leftover.

### CODE REVIEW

**Comment by vspetrov** [comment]
@manjugv @Sergei-Lebedev @bureddy it is ready for review now

File: src/components/mc/cpu/mc_cpu_reduce.h [hunk 1]
```diff
@@ -9,77 +9,158 @@
 
 #include "utils/ucc_math.h"
 
-#define DO_DT_REDUCE_WITH_OP(s1, s2, d, count, OP) do {                        \
+#define OP_1(_s1, _s2, _i, _sc, _OP) _OP(_s1[_i], _s2[_i])
```
Comment (Sergei-Lebedev) [inline src/components/mc/cpu/mc_cpu_reduce.h:12]: Minor, manually substituting OP reduces mc_cpu compilation time from 68 sec. to 58 sec.

```cpp
#define OP_1(_s1, _s2, _i, _sc, _OP) _OP(_s1[_i], _s2[_i])
#define OP_2(_s1, _s2, _i, _sc, _OP) _OP(_OP(_s1[_i], _s2[_i]),  _s2[_i + 1 * _sc])
#define OP_3(_s1, _s2, _i, _sc, _OP) _OP(_OP(_s1[_i], _s2[_i]),  _OP(_s2[_i + 1 * _sc],  _s2[_i + 2 * _sc]))
#define OP_4(_s1, _s2, _i, _sc, _OP) _OP(_OP(_OP(_s1[_i], _s2[_i]),  _OP(_s2[_i + 1 * _sc],  _s2[_i + 2 * _sc])), _s2[_i + 3 * _sc])
#define OP_5(_s1, _s2, _i, _sc, _OP) _OP(_OP(_OP(_s1[_i], _s2[_i]),  _OP(_s2[_i + 1 * _sc],  _s2[_i + 2 * _sc])), _OP(_s2[_i + 3 * _sc], _s2[_i + 4 * _sc]))
#define OP_6(_s1, _s2, _i, _sc, _OP) _OP(_OP(_OP(_s1[_i], _s2[_i]),  _OP(_s2[_i + 1 * _sc],  _s2[_i + 2 * _sc])), _OP(_OP(_s2[_i + 3 * _sc], _s2[_i + 4 * _sc]),_s2[_i + 5 * _sc]))
#define OP_7(_s1, _s2, _i, _sc, _OP) _OP(_OP(_OP(_OP(_s1[_i], _s2[_i]),  _OP(_s2[_i + 1 * _sc],  _s2[_i + 2 * _sc])), _OP(_OP(_s2[_i + 3 * _sc], _s2[_i + 4 * _sc]),_s2[_i + 5 * _sc])), _s2[_i + 6 * _sc])
```

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/mc/cpu/mc_cpu_reduce.h (hunk 1 above)
Comment (vspetrov) [reply src/components/mc/cpu/mc_cpu_reduce.h]: After another round of hesitations i split reductions into several independent functions and put them into different .c files. The compilation now 4 secs. The interfaces of the MC is kept the same (it 1 reduce_multi function). @Sergei-Lebedev plz have another look

**Comment by artemry-nv** [comment]
bot:retest

**Review by manjugv — APPROVED** [review]

**Pull request merged**